### PR TITLE
feat(formatter): add `delimiterSpacing` option

### LINKED ISSUE #4607: 📎 Implement formatting option to include space inside parens, braces, brackets, etc.
### Description

Add formatting option to include spacing inside parentheses, square brackets, curly braces, and angle brackets.

For example, this input:

```ts
function id<TypeArg>(x: TypeArg): TypeArg {
	return x;
}
const arr = [1, 2, 3];
if (id(true)) {
	// …
}
```

Would be formatted as follows with a "space inside" option enabled:

```ts
function id< TypeArg > ( x: TypeArg ): TypeArg {
	return x;
}
const arr = [ 1, 2, 3 ];
if ( id( true ) ) {
	// …
}
```

See https://github.com/biomejs/biome/discussions/2360#discussioncomment-10690814.

### PR DESCRIPTION
**Pull request by luisherranz**
_Huge kudos to @orballo for doing most of the work in discovering which parts of the Biome formatter needed to be changed to add this option in his proof of concept: https://github.com/orballo/biome/pull/1_

## Summary

This PR adds a new `delimiterSpacing` formatter option that inserts spaces inside delimiters (parentheses, square brackets, TS angle brackets, and JSX curly braces).

**Closes #4607**

When enabled, this input:

```ts
function id<TypeArg>(x: TypeArg): TypeArg {
  return x;
}
const arr = [1, 2, 3];
if (id(true)) {
  // …
}
```

Is formatted as:

```ts
function id< TypeArg >( x: TypeArg ): TypeArg {
  return x;
}
const arr = [ 1, 2, 3 ];
if ( id( true ) ) {
  // …
}
```

The option is available for:
- **JavaScript/TypeScript**: Parentheses `()`, square brackets `[]`, TS angle brackets `<>`, and JSX curly braces `{}`
- **JSON**: Square brackets `[]`
- **CSS**: Parentheses `()` and square brackets `[]`

Configuration example:

```json
{
  "formatter": {
    "delimiterSpacing": true
  }
}
```

Or per-language:

```json
{
  "javascript": {
    "formatter": {
      "delimiterSpacing": true
    }
  }
}
```

See https://github.com/biomejs/biome/discussions/2360#discussioncomment-10690814.

## AI usage

As the [Gutenberg](https://github.com/wordpress/gutenberg) project is using a [non-official fork of Prettier](https://github.com/WordPress/gutenberg/blob/e2d8385f01f5995cf69bdfb40b8b4d9c8015eed0/package.json#L102), I run [@orballo's Biome version](https://github.com/orballo/biome/pull/1) in the Gutenberg repository and used AI to extract all the tests and edge cases.  Then the AI assisted me in converting the changes made by @orballo's proof of concept to changes in the current formatter.

## Test Plan

- [x] Added snapshot tests for all supported languages
- [x] Updated existing formatter snapshots to include the new option

## Website

Website docs and playground integration PR in this website repo PR:

- https://github.com/biomejs/website/pull/3910

### CODE REVIEW

File: crates/biome_js_formatter/src/js/assignments/array_assignment_pattern.rs [hunk 1]
```diff
@@ -23,7 +23,14 @@ impl FormatNodeRule<JsArrayAssignmentPattern> for FormatJsArrayAssignmentPattern
...
         } else {
-            write!(f, [group(&soft_block_indent(&elements.format()))])?;
+            let should_insert_space_around_brackets = f.options().delimiter_spacing().value();
+            write!(
+                f,
+                [group(&soft_block_indent_with_maybe_space(
+                    &elements.format(),
+                    should_insert_space_around_brackets
+                ))]
+            )?;
```
[the remark this answers is not in the document]
Comment (luisherranz) [reply crates/biome_js_formatter/src/js/assignments/array_assignment_pattern.rs:33]: Empty delimiters don't need a space for readability.

File: crates/biome_js_formatter/src/js/expressions/unary_expression.rs [hunk 2]
```diff
@@ -1,8 +1,8 @@
 use crate::prelude::*;
 
 use biome_formatter::{format_args, write};
-use biome_js_syntax::JsUnaryExpression;
 use biome_js_syntax::parentheses::NeedsParentheses;
+use biome_js_syntax::{AnyJsExpression, JsUnaryExpression};
```
[the remark this answers is not in the document]
Comment (luisherranz) [reply crates/biome_js_formatter/src/js/expressions/unary_expression.rs:5]: Coding standards that prioritize accessibility tend to include a space after the negation operator to improve readability so the space after `!` is intentional.

File: crates/biome_js_formatter/src/js/expressions/arrow_function_expression.rs [hunk 3]
```diff
@@ -123,7 +124,10 @@ impl FormatNodeRule<JsArrowFunctionExpression> for FormatJsArrowFunctionExpressi
                                 group(&format_args![
                                     space(),
                                     token("("),
-                                    soft_block_indent(&format_body),
+                                    soft_block_indent_with_maybe_space(
+                                        &format_body,
+                                        should_insert_space
+                                    ),
                                     token(")")
                                 ])
```
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/biome_js_formatter/src/js/expressions/arrow_function_expression.rs:132]: @coderabbitai could you suggest a test that could fail because of this?

[the remark this answers is not in the document]
Comment (ematipico) [reply crates/biome_js_formatter/src/js/expressions/arrow_function_expression.rs:132]: @luisherranz here you go

File: crates/biome_js_formatter/src/js/classes/setter_class_member.rs [hunk 4]
```diff
@@ -20,21 +20,44 @@ impl FormatNodeRule<JsSetterClassMember> for FormatJsSetterClassMember {
...
+                    l_paren_token.format(),
+                    space(),
+                    parameter.format(),
+                    comma_token.format(),
+                    space(),
+                    r_paren_token.format(),
+                    space(),
+                    body.format(),
+                ]
+            ]
```
[the remark this answers is not in the document]
Comment (luisherranz) [reply crates/biome_js_formatter/src/js/classes/setter_class_member.rs:43]: Added a test that shows this is wrong in https://github.com/biomejs/biome/pull/8966/commits/a91e20baf2d247155386d1fc3c0471e45660d0c1.

File: crates/biome_js_formatter/src/js/expressions/arrow_function_expression.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (luisherranz) [reply crates/biome_js_formatter/src/js/expressions/arrow_function_expression.rs:132]: Thanks @ematipico. I'll take a look at this and the rest of the comments later tonight 🙂

[the remark this answers is not in the document]
Comment (luisherranz) [reply crates/biome_js_formatter/src/js/expressions/arrow_function_expression.rs:132]: @coderabbitai, this is wrong. 

`soft_block_indent_with_maybe_space` delegates to `soft_space_or_block_indent` (when `true`), which uses `IndentMode::SoftSpace`. The `BlockIndent` implementation in `crates/biome_formatter/src/builders.rs` writes `soft_line_break_or_space()` on **both** sides of the content (before at lines 1573-1574, after at line 1595). When the group fits on one line, both resolve to a space character, producing symmetric output.

The existing snapshot already proves this, see `arrow_functions.js.snap` line 321:
```
const foo1 = ( a ) => ( a, b );
```

Both opening and closing delimiters have spaces.

File: crates/biome_js_formatter/src/js/statements/while_statement.rs [hunk 5]
```diff
@@ -18,13 +18,18 @@ impl FormatNodeRule<JsWhileStatement> for FormatJsWhileStatement {
...
             [group(&format_args![
                 while_token.format(),
                 space(),
                 l_paren_token.format(),
-                group(&soft_block_indent(&test.format())),
+                group(&soft_block_indent_with_maybe_space(
+                    &test.format(),
+                    should_insert_space
+                )),
                 r_paren_token.format(),
```
[the remark this answers is not in the document]
Comment (luisherranz) [reply crates/biome_js_formatter/src/js/statements/while_statement.rs:33]: @coderabbitai, this is wrong. 


`soft_block_indent_with_maybe_space` delegates to `soft_space_or_block_indent` (when `true`), which uses `IndentMode::SoftSpace`. The `BlockIndent` implementation in `crates/biome_formatter/src/builders.rs` writes `soft_line_break_or_space()` on **both** sides of the content, before and after. When the group fits on one line, both resolve to a space character, producing symmetric output.

The existing snapshot already proves this, see `loops.js.snap`:
```
while ( a ) {}
do {} while ( a );
```

Both opening and closing delimiters have spaces.

File: crates/biome_js_formatter/src/jsx/auxiliary/spread_child.rs [hunk 6]
```diff
@@ -51,7 +52,21 @@ impl FormatNodeRule<JsxSpreadChild> for FormatJsxSpreadChild {
...
+                        &format_inner,
+                        true
+                    ))]
+                )?;
+            } else {
+                // Without delimiter spacing, always break to multiple lines
+                write!(f, [soft_block_indent(&format_inner)])?;
+            }
+        } else if delimiter_spacing {
+            write!(f, [space(), format_inner, space()])?;
```
[the remark this answers is not in the document]
Comment (luisherranz) [reply crates/biome_js_formatter/src/jsx/auxiliary/spread_child.rs:69]: @coderabbitai, this is wrong. 

`soft_block_indent_with_maybe_space` delegates to `soft_space_or_block_indent` (when `true`), which uses `IndentMode::SoftSpace`. The `BlockIndent` implementation in `crates/biome_formatter/src/builders.rs` writes `soft_line_break_or_space()` on **both** sides of the content, before and after. When the group fits on one line, both resolve to a space character, producing symmetric output.

The existing snapshot already proves this — see `jsx_expression_children.jsx.snap`:
```
<Foo>{ ...a /* b */ }</Foo>;
<Foo>{ /* a */ ...b /* c */ }</Foo>;
```

Both opening and closing delimiters have spaces, even with comments present.

File: crates/biome_js_formatter/tests/specs/js/module/delimiter-spacing/setters.js [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+// Boundary test (fits): 78+2=80 chars with delimiter spacing
+class Foo19 {
+	set a({ a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, uv }) {}
+}
+
+// Boundary test (breaks): 79+2=81 chars with delimiter spacing
+class Foo20 {
+	set a({ a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, uvw }) {}
+}
```
[the remark this answers is not in the document]
Comment (luisherranz) [reply crates/biome_js_formatter/tests/specs/js/module/delimiter-spacing/setters.js:137]: @coderabbitai, not sure what you are measuring but `Foo19` and `Foo20` are correct in the snapshot:

When `delimiterSpacing` is false ([file](https://github.com/luisherranz/biome/blob/a91e20baf2d247155386d1fc3c0471e45660d0c1/crates/biome_js_formatter/tests/specs/js/module/delimiter-spacing/setters.js.snap?plain=1#L402-L410)):

```js
// Boundary test (fits): 78+2=80 chars with delimiter spacing
class Foo19 {
	set a({ a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, uv }) {}
}

// Boundary test (breaks): 79+2=81 chars with delimiter spacing
class Foo20 {
	set a({ a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, uvw }) {}
}
```
When `delimiterSpacing` is true ([file](https://github.com/luisherranz/biome/blob/a91e20baf2d247155386d1fc3c0471e45660d0c1/crates/biome_js_formatter/tests/specs/js/module/delimiter-spacing/setters.js.snap?plain=1#L663-L693)):

```js
// Boundary test (fits): 78+2=80 chars with delimiter spacing
class Foo19 {
	set a( { a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, uv } ) {}
}

// Boundary test (breaks): 79+2=81 chars with delimiter spacing
class Foo20 {
	set a( {
		a,
		b,
		c,
		d,
		e,
		f,
		g,
		h,
		i,
		j,
		k,
		l,
		m,
		n,
		o,
		p,
		q,
		r,
		s,
		t,
		uvw,
	} ) {}
}
```

**Comment by dyc3** [comment]
It would be good to also update some of the rule metadata in `crates/biome_cli/src/execute/migrate/unsupported_rules.rs`, like for `array-bracket-spacing` or `space-in-parens`. There are probably some others.

**Comment by luisherranz** [comment]
Sure, I'll try to address that as well.

**Comment by luisherranz** [comment]
@dyc3, there you go: https://github.com/biomejs/biome/pull/8966/commits/bea1683c62063a14d26afcd125ef6e3a37a5dd28

Let me know if anything else is needed.

**Comment by luisherranz** [comment]
> update the UnsupportedRule for
EslintStylistic("type-generic-spacing") to use
FormatterOption("delimiterSpacing") so it aligns with other delimiter rules

@coderabbitai, [`type-generic-spacing`](https://eslint.style/rules/type-generic-spacing) is not affected by `delimiterSpacing` so it's correct as it is.

File: crates/biome_cli/src/execute/migrate/unsupported_rules.rs [hunk 8]
```diff
@@ -85,15 +91,24 @@ pub const UNSUPPORTED_RULES: &[UnsupportedRule] = &[
...
     UnsupportedRule(Eslint("space-return-throw-case"), FormatterCovers),
     UnsupportedRule(Eslint("space-unary-ops"), FormatterCovers),
     UnsupportedRule(Eslint("space-unary-word-ops"), FormatterCovers),
     UnsupportedRule(Eslint("spaced-comment"), Stylistic),
     UnsupportedRule(Eslint("switch-colon-spacing"), FormatterCovers),
-    UnsupportedRule(Eslint("template-curly-spacing"), FormatterCovers),
+    UnsupportedRule(
+        Eslint("template-curly-spacing"),
+        FormatterOption("delimiterSpacing"),
+    ),
```
[the remark this answers is not in the document]
Comment (luisherranz) [reply crates/biome_cli/src/execute/migrate/unsupported_rules.rs:111]: Fixed in https://github.com/biomejs/biome/pull/8966/commits/ef5fe9c9fcd8e20bd1622e8dddf2ecfcdc79b0b1.

File: crates/biome_css_formatter/tests/specs/css/delimiter_spacing/container.css.snap [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+
+/* Style queries */
+@container style( --theme: dark ) {
+	.themed {
+		background: black;
+		color: white;
+	}
+}
+
+@container style( --responsive: true ) {
```
Comment (ematipico) [inline crates/biome_css_formatter/tests/specs/css/delimiter_spacing/container.css.snap:240]: `style` here should have a space

File: crates/biome_css_formatter/tests/specs/css/delimiter_spacing/functions.css.snap [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+div {
+	color: rgba( 255, 255, 255, 1 );
+	color: rgb( 100, 200, 50 );
+	color: hsl( 120, 50%, 50% );
+	color: hsla( 120, 50%, 50%, 0.5 );
+	color: hwb( 120 20% 30% );
+	color: lab( 50% 40 59.5 );
+	color: lch( 50% 59.5 40 );
+	color: oklch( 0.7 0.2 30 );
+	color: oklab( 0.5 0.1 0.2 );
```
Comment (ematipico) [inline crates/biome_css_formatter/tests/specs/css/delimiter_spacing/functions.css.snap:174]: This seems intentional, however `container` had a space before the `(`. Which is it? Should we aim for consistency?

**Comment by ematipico** [comment]
Thank you for this!

So, this is a massive PR and I wonder why this wasn't split in the first place. It will take us a while to review it, and since we're also working on SCSS, it will take you a lot of work to keep the PR up to date.

Wouldn't be better to implement this by language?

As for docs, I gave it a glance, I already started seeing a problem: they are very generic and repeated. We should aim for preciseness, for users **and us**. I want to understand the criteria of option, so that if I need to triage a bug, I know what to do.

For example `function f()` has delimiters, do we place a space between `f` and `(`. We should address that.

File: crates/biome_css_formatter/tests/specs/css/delimiter_spacing/container.css.snap (hunk 9 above)
Comment (luisherranz) [reply crates/biome_css_formatter/tests/specs/css/delimiter_spacing/container.css.snap:240]: The `delimiter_spacing` option only adds spaces _inside_ delimiters (after `(` and before `)`), not before the opening delimiter. So `style( --responsive: true )` is the intended output, the same way `rgba( 255, 255, 255, 1 )` doesn't get a space before `(` either.

I've updated the docs in https://github.com/biomejs/biome/pull/8966/commits/8ea460f0a2d473e03a52600d7bdd2372a15ae0fa to make this explicit: "Spaces are not added before the opening delimiter", but if you think it needs more clarification, let me know.

File: crates/biome_css_formatter/tests/specs/css/delimiter_spacing/functions.css.snap (hunk 10 above)
Comment (luisherranz) [reply crates/biome_css_formatter/tests/specs/css/delimiter_spacing/functions.css.snap:174]: The space before `(` in `@container ( min-width: 400px )` is part of the CSS at-rule syntax, it's always there, even with `delimiter_spacing: false`. On the other hand, `style(` and `rgba(` are function-call syntax where there's never a space before `(`. The `delimiter_spacing` option only adds spaces _inside_ delimiters, not before them. So this is consistent; it's just two different syntactic constructs.

**Comment by luisherranz** [comment]
Thanks for taking a look, @ematipico! I really appreciate it.

> As for docs, I gave it a glance, I already started seeing a problem: they are very generic and repeated. We should aim for preciseness, for users and us. I want to understand the criteria of option, so that if I need to triage a bug, I know what to do.

I've improved the docs in [this commit](https://github.com/biomejs/biome/pull/8966/commits/8ea460f0a2d473e03a52600d7bdd2372a15ae0fa), but if you still think they are not accurate enough, please give me an example of how you think they should be and I'll try to improve them. Once we have them right, I'll also update the PR on the website.

> since we're also working on SCSS, it will take you a lot of work to keep the PR up to date.
> Wouldn't be better to implement this by language?

I can split the PR into three or add the option to the SCSS part and keep it up to date, whatever you folks prefer.

**Comment by ematipico** [comment]
> I can split the PR into three or add the option to the SCSS part and keep it up to date, whatever you folks prefer.

It would be great if you could split up the PR in multiple PRs (as you see fit). That would help. ,For example I saw a bug but I lost the file because there are too many. 

I recall that in `foo<T>(), <T>` doesn't get the spaces, but I seem to remember that we document that the spaces are applied to angle brackets `<>`

**Comment by luisherranz** [comment]
> It would be great if you could split up the PR in multiple PRs (as you see fit)

I've split the PR into all these branches, but I've just realized that they can't be opened with the correct base if some branches originate from others when all of them are coming from a fork. 

What do you want me to do? Should I open them all against `upstream:next`?

- `feat/delimiter-spacing-core`: shared infrastructure ([branch against next](https://github.com/luisherranz/biome/compare/next...luisherranz:biome:feat/delimiter-spacing-core))
  - `feat/delimiter-spacing-js`: JavaScript implementation ([branch against feat/delimiter-spacing-core](https://github.com/luisherranz/biome/compare/feat/delimiter-spacing-core...luisherranz:biome:feat/delimiter-spacing-js))
    - `feat/delimiter-spacing-jsx`: JSX implementation ([branch against feat/delimiter-spacing-js](https://github.com/luisherranz/biome/compare/feat/delimiter-spacing-js...luisherranz:biome:feat/delimiter-spacing-jsx))
    - `feat/delimiter-spacing-ts`: TypeScript implementation ([branch against feat/delimiter-spacing-js](https://github.com/luisherranz/biome/compare/feat/delimiter-spacing-js...luisherranz:biome:feat/delimiter-spacing-ts))
  - `feat/delimiter-spacing-css`: CSS implementation ([branch against feat/delimiter-spacing-core](https://github.com/luisherranz/biome/compare/feat/delimiter-spacing-core...luisherranz:biome:feat/delimiter-spacing-css))
  - `feat/delimiter-spacing-json`: JSON implementation ([branch against feat/delimiter-spacing-core](https://github.com/luisherranz/biome/compare/feat/delimiter-spacing-core...luisherranz:biome:feat/delimiter-spacing-json))
  -  `feat/delimiter-spacing-eslint`: ESLint migration mappings ([branch against feat/delimiter-spacing-core](https://github.com/luisherranz/biome/compare/feat/delimiter-spacing-core...luisherranz:biome:feat/delimiter-spacing-eslint))

All branches target `next`. ESLint, JS, CSS, and JSON depend on core. JSX and TS depend on JS.

**Comment by ematipico** [comment]
> What do you want me to do? Should I open them all against upstream:next?

Yeah, that sounds good! HEADS UP! I just merged a few things around the formatter testing infrastructure. It's now less noisy, and the snapshots are different. You should see different outputs from the snapshots, and less changes. Thankfully, the PR should be more bearable.

I'll help you to bring the PRs to the finish line

**Comment by luisherranz** [comment]
> I just merged a few things around the formatter testing infrastructure. It's now less noisy, and the snapshots are different. You should see different outputs from the snapshots, and less changes. Thankfully, the PR should be more bearable.

Looks great!

However, I see some errors in `next` right now. One is in `JsonAnalyzeServices`, which is missing `project_layout` and preventing the build. Once that's fixed, there are 26 HTML snapshots from Astro, Vue, and Svelte that are not up to date.

Do you want me to open a PR to fix those things before opening mines, or could you let me know when that's resolved? Thank you very much.

**Comment by ematipico** [comment]
I can fix those, so you can focus on your stuff. If they prevent you from moving forward, feel free to fix them on your branch (the formatting one). I'll help to update the PRs once the fixes are in next

**Comment by luisherranz** [comment]
Great, thank you very much.

Let me know when `next` is free of issues and I'll rebase my branches and open all the PRs with all the updated snapshots.

**Comment by ematipico** [comment]
@luisherranz the next branch should be good to go

**Comment by luisherranz** [comment]
Thanks @ematipico!

I've already opened all the PRs against `next`. They're still in draft because I wanted to take one last look at them tomorrow, but as soon as I review them, I'll mark them as ready for review.

- #9718 (core)
  - #9719 (JS)
    - #9720 (JSX, depends on JS)
      - #9721 (TypeScript, depends on JS and JSX)
  - #9722 (CSS)
  - #9723 (JSON)
  - #9724 (ESLint migration)

**Comment by luisherranz** [comment]
Superseed by https://github.com/biomejs/biome/pull/9718.

**Comment by luisherranz** [comment]
The main branch adding the option and all the JavaScript branches are already ready for review. 

Tomorrow or the day after, I'll continue with CSS, JSON, and ESLint.
